wa/workloads/speedometer/__init__.py

```diff
@@ -12,57 +12,363 @@
+
+            time.sleep(sleep_period_s)
+
+    def read_score(self, context):
+        xml_dump = os.path.join(self.target.working_directory, "speedometer_dump.xml")
+
+        self.target.execute("uiautomator dump {}".format(xml_dump), as_root=True)
+        self.target.pull(xml_dump, context.output_directory)
+
+        with open(
+            os.path.join(context.output_directory, "speedometer_dump.xml"), "rb"
+        ) as fh:
+            dump = fh.read().decode("utf-8")
+        match = self.regex.search(dump)
+        result = None
+        if match:
+            result = float(match.group(1))
+
+        return result
 
     def update_output(self, context):
         super(Speedometer, self).update_output(context)
-        result = None
-        logcat_file = context.get_artifact_path('logcat')
-        with open(logcat_file, errors='replace') as fh:
-            for line in fh:
-                match = self.regex.search(line)
-                if match:
-                    result = float(match.group(1))
-
-        if result is not None:
-            context.add_metric('Speedometer Score', result, 'Runs per minute', lower_is_better=False)
-        else:
-            raise WorkloadError("The Speedometer workload has failed. No score was obtainable.")
+
+        score_read = False
+        iterations = 0
+        while not score_read:
+            score = self.read_score(context)
+
+            if score is not None:
+                context.add_metric(
+                    "Speedometer Score", score, "Runs per minute", lower_is_better=False
+                )
+                score_read = True
+            else:
+                if iterations >= 10:
+                    raise WorkloadError(
+                        "The Speedometer workload has failed. No score was obtainable."
+                    )
+                else:
+                    # Sleep and retry...
+                    time.sleep(2)
+                    iterations += 1
+
+    def teardown(self, context):
+        super(Speedometer, self).teardown(context)
+
+        # Shutdown the locally hosted version of Speedometer
+        self.archive_server.stop(self.target)
+
+        # Cleanup temporary files from the output directory.
+        files_to_delete = [
+            self.document_root,
+            os.path.join(context.output_directory, "speedometer_dump.xml"),
+            os.path.join(context.output_directory, self.preferences_xml),
+            os.path.join(context.output_directory, self.preferences_xml + ".new"),
+        ]
+        for file_to_delete in files_to_delete:
+            if os.path.exists(file_to_delete):
+                subprocess.check_output("rm -r {}".format(file_to_delete), shell=True)
+
+        # The browser's processes can stick around and have minor impact on
+        # other performance sensitive workloads, so make sure we clean up.
+        self.target.execute("am force-stop {}".format(self.chrome_package))
+
```

Comment:
Add a `finalize()` to clean up the workload files deployed to the target if `self.cleanup_assets` is `True`